lib/model/query/entities.js

```diff
@@ -106,16 +109,36 @@ const _processSubmissionDef = (submissionDefId, submissionId) => async ({ Datase
   // if the form def is about datasets and entities.
   if (fields.length === 0)
     return null;
-  const partial = await Entity.fromSubmissionXml(submissionDef.xml, fields);
-  // If no data was returned, e.g. if create flag was not set, then don't continue
-  if (!partial)
+
+  const entityData = await parseSubmissionXml(fields, submissionDef.xml);
+
+  // If dataset requires approval on submission to create an entity and this event is not
+  // an approval event, then don't create an entity
+  //
+  // We can't simply use submissionDefId to trace back to dataset and find out if approval
+  // is required because in future there can be multiple entities in a single submission.
+  // So we have to rely on dataset name from the xml to fetch the dataset.approvalRequired flag.
```

Comment:
If we're fetching the dataset from the name in the submission xml here, maybe we could just pass the dataset to `Entities.createNew` and not have to do the lookup in `_getDataset`, thus simplifying that function or not even needing it because we'll always have the dataset id. It just needs to be set in `partial.datasetId`.